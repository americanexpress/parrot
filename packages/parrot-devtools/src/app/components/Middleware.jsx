/*
 * Copyright (c) 2018 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

class Middleware extends Component {
  // eslint-disable-next-line react/state-in-constructor
  state = {
    loading: true,
  };

  // eslint-disable-next-line react/no-deprecated
  componentWillMount() {
    // eslint-disable-next-line react/destructuring-assignment
    this.loadScenarios(this.props.url);
  }

  // eslint-disable-next-line react/no-deprecated
  componentWillReceiveProps({ url: nextUrl }) {
    // eslint-disable-next-line react/destructuring-assignment
    if (this.props.url !== nextUrl) {
      this.loadScenarios(nextUrl);
    }
  }

  setScenario = async scenario => {
    const { url } = this.props;
    try {
      await fetch(`${url}/parrot/scenario`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          scenario,
        }),
      });
      this.setState({ scenario });

      if (chrome) {
        chrome.devtools.inspectedWindow.reload(null);
      }
    } catch (e) {
      // do nothing
    }
  };

  loadScenarios = async url => {
    this.setState({ loading: true });
    try {
      const scenariosResponse = await fetch(`${url}/parrot/scenarios`);
      const scenarios = await scenariosResponse.json();
      const scenarioResponse = await fetch(`${url}/parrot/scenario`);
      const scenario = await scenarioResponse.json();
      this.setState({
        scenarios,
        scenario,
        loading: false,
      });
    } catch (e) {
      this.setState({ loading: false });
    }
  };

  render() {
    const { render } = this.props;
    const { scenarios, scenario, loading } = this.state;
    return render(
      { scenarios, scenario, setScenario: this.setScenario, loadScenarios: this.loadScenarios },
      loading
    );
  }
}

Middleware.propTypes = {
  url: PropTypes.string.isRequired,
  render: PropTypes.func.isRequired,
};

export const mapStateToProps = ({ url }) => ({ url });

export default connect(mapStateToProps)(Middleware);
