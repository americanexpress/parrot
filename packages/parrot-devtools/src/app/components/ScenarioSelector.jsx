/*
 * Copyright (c) 2018 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Scrollable } from './styled';

const ScenarioSelector = ({
  data: { scenarios, scenario, setScenario, loadScenarios },
  loading,
}) => {
  if (loading) {
    return (
      <Grid>
        <div className="progress-circle progress-indeterminate margin-auto" />
      </Grid>
    );
  }

  if (!scenarios || !scenario) {
    return (
      <Scrollable>
        <div className="alert alert-warn alert-dialog alert-lg" role="alert">
          <span className="alert-icon dls-icon-warning icon-lg" />
          <h1 className="heading-4 alert-header">Failed to Load Scenarios</h1>
          <p className="alert-subtitle">
            {/* eslint-disable-next-line jsx-a11y/anchor-is-valid, jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */}
            Try <a onClick={loadScenarios}>refreshing</a> or updating the middleware URL in
            settings.
          </p>
        </div>
      </Scrollable>
    );
  }

  return (
    <Scrollable className="border">
      <ul className="nav-menu">
        {scenarios.map(({ name }) => (
          <li className="nav-item" key={name}>
            {/* eslint-disable-next-line jsx-a11y/anchor-is-valid, jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */}
            <a
              onClick={() => setScenario(name)}
              className="nav-link fluid"
              aria-selected={name === scenario}
            >
              {name}
            </a>
          </li>
        ))}
      </ul>
    </Scrollable>
  );
};

ScenarioSelector.propTypes = {
  data: PropTypes.shape({
    scenarios: PropTypes.array,
    scenario: PropTypes.string,
    setScenario: PropTypes.func.isRequired,
    loadScenarios: PropTypes.func.isRequired,
  }).isRequired,
  loading: PropTypes.bool.isRequired,
};

export default ScenarioSelector;
