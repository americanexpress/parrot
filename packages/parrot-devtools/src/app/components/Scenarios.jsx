/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';

import { Grid, Scrollable } from './styled';

function ScenariosMiddleware({ scenariosData }) {
  const { loading, filteredScenarios, scenario, setScenario, loadScenarios } = scenariosData;

  if (loading) {
    return (
      <Grid>
        <div className="progress-circle progress-indeterminate margin-auto" />
      </Grid>
    );
  }

  if (!filteredScenarios || !scenario) {
    return (
      <Scrollable>
        <div className="alert alert-warn alert-dialog alert-lg" role="alert">
          <span className="alert-icon dls-icon-warning icon-lg" />
          <h1 className="heading-4 alert-header">Failed to Load Scenarios</h1>
          <p className="alert-subtitle">
            <span>Try </span>
            <button type="button" onClick={loadScenarios}>
              refreshing
            </button>
            <span> or updating the middleware URL in settings.</span>
          </p>
        </div>
      </Scrollable>
    );
  }

  return (
    <Scrollable className="border">
      <ul className="nav-menu">
        {filteredScenarios.map(({ name }) => (
          <li className="nav-item" key={name}>
            <button
              type="button"
              className={`nav-link btn-secondary text-align-left fluid ${
                name === scenario ? '' : 'border-0'
              }`}
              aria-pressed={name === scenario}
              onClick={() => setScenario(name)}
            >
              {name}
            </button>
          </li>
        ))}
      </ul>
    </Scrollable>
  );
}

ScenariosMiddleware.propTypes = {
  scenariosData: PropTypes.shape({
    loading: PropTypes.bool,
    filteredScenarios: PropTypes.array,
    scenario: PropTypes.string,
    setScenario: PropTypes.func,
    loadScenarios: PropTypes.func,
  }).isRequired,
};

export default ScenariosMiddleware;
