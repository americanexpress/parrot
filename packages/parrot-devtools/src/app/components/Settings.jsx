/*
 * Copyright (c) 2018 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { setUrl as setUrlAction } from '../ducks';
import { Scrollable } from './styled';

const Settings = ({ url, setUrl }) => (
  <Scrollable>
    <div className="pad-1">
      <h1 className="heading-5 pad-b">Settings</h1>
      <div className="form-group">
        {/* eslint-disable-next-line jsx-a11y/label-has-associated-control */}
        <label htmlFor="url">Middleware URL</label>
        <input
          onBlur={({ target: { value } }) => setUrl(value)}
          defaultValue={url}
          id="url"
          className="form-control"
        />
      </div>
    </div>
  </Scrollable>
);

Settings.propTypes = {
  url: PropTypes.string.isRequired,
  setUrl: PropTypes.func.isRequired,
};

export const mapStateToProps = ({ url }) => ({ url });

export const mapDispatchToProps = dispatch => ({
  setUrl: url => dispatch(setUrlAction(url)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Settings);
