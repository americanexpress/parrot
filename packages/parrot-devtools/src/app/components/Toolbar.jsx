/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { SearchBar, ClearButton } from './styled';

const Toolbar = ({ scenariosData }) => {
  const { filterValue, setFilterValue } = scenariosData;
  const handleFilterChange = e => {
    scenariosData.setFilterValue(e.target.value.toLowerCase());
  };
  return (
    <div className="pad-1-b flex flex-justify-center flex-align-items-center">
      <SearchBar
        value={filterValue}
        onChange={handleFilterChange}
        placeholder="filter scenarios..."
      />
      <ClearButton className="dls-glyph-cancel-circle" onClick={() => setFilterValue('')} />
    </div>
  );
};

export default Toolbar;

Toolbar.propTypes = {
  scenariosData: PropTypes.shape({
    filterValue: PropTypes.string,
    setFilterValue: PropTypes.func,
  }).isRequired,
};
