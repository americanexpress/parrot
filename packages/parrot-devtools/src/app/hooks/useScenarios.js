/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';

export default function useScenarios(url, initialScenario) {
  const [loading, setLoading] = React.useState(false);
  const [scenario, setScenario] = React.useState('');
  const [scenarios, setScenarios] = React.useState([]);

  const methods = React.useMemo(
    () => ({
      loadScenarios: async () => {
        setLoading(true);
        try {
          const scenariosResponse = await fetch(`${url}/parrot/scenarios`);
          const scenarioResponse = await fetch(`${url}/parrot/scenario`);
          // eslint-disable-next-line prettier/prettier
          setScenario( await scenarioResponse.json() );
          // eslint-disable-next-line prettier/prettier
          setScenarios( await scenariosResponse.json() );
        } catch (e) {
          // do nothing
        }
        setLoading(false);
      },
      setScenario: async selectedScenario => {
        try {
          await fetch(`${url}/parrot/scenario`, {
            method: 'POST',
            headers: {
              'Content-Type': 'application/json',
            },
            body: JSON.stringify({
              scenario: selectedScenario,
            }),
          });

          setScenario(selectedScenario);

          if ('chrome' in window) {
            chrome.devtools.inspectedWindow.reload(null);
          }
        } catch (e) {
          // do nothing
        }
      },
    }),
    [url]
  );

  React.useLayoutEffect(() => {
    if (initialScenario) methods.setScenario(initialScenario);
    methods.loadScenarios();
  }, [url, initialScenario]);

  const [filteredScenarios, setFilteredScenarios] = React.useState(scenarios);
  const [filterValue, setFilterValue] = React.useState('');

  React.useEffect(() => {
    if (filterValue === '') {
      setFilteredScenarios(scenarios);
    } else {
      setFilteredScenarios(
        scenarios.filter(({ name }) => name.toLowerCase().includes(filterValue))
      );
    }
  }, [filterValue, scenarios]);

  return {
    loading,
    scenario,
    filteredScenarios,
    filterValue,
    setFilterValue,
    ...methods,
  };
}
