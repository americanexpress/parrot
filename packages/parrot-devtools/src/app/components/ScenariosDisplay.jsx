/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Content } from './styled';
import Scenarios from './Scenarios';
import Toolbar from './Toolbar';
import { useScenarios } from '../hooks';

function ScenariosDisplay({ url }) {
  const scenariosData = useScenarios(url);
  return (
    <Fragment>
      <Toolbar scenariosData={scenariosData} />
      <Content>
        <Scenarios scenariosData={scenariosData} />
      </Content>
    </Fragment>
  );
}

ScenariosDisplay.propTypes = {
  url: PropTypes.string.isRequired,
};

export const mapStateToProps = ({ url }) => ({ url });

export const ComponentUnderTest = ScenariosDisplay;
export default connect(mapStateToProps)(ScenariosDisplay);
