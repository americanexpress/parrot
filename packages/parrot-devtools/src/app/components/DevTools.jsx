/*
 * Copyright (c) 2018 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

import React from 'react';

import parrotLogo from '../../assets/img/parrot_128x.png';

import { useDevTools } from '../hooks';

import { Container, Content, Navigation, Logo } from './styled';
import Settings from './Settings';
import ScenariosDisplay from './ScenariosDisplay';

export default function DevTools() {
  const { showSettings, toggleSettings } = useDevTools();

  return (
    <Container className="pad-1 dls-accent-white-01-bg">
      <Content className={showSettings ? '' : 'display-none'}>
        <Settings />
      </Content>
      <ScenariosDisplay className={showSettings ? 'display-none' : ''} />
      <Navigation className="pad-1-t flex flex-align-center flex-justify-between">
        <Logo src={parrotLogo} />
        {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
        <button
          type="button"
          onClick={toggleSettings}
          className={`btn-icon btn-inline ${
            showSettings ? 'btn-primary' : 'btn-tertiary'
          } dls-icon-setting`}
          title={showSettings ? 'Hide settings' : 'Show settings'}
        />
      </Navigation>
    </Container>
  );
}
