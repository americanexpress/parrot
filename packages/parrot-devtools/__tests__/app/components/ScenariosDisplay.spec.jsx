/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { shallow } from 'enzyme';

import { ComponentUnderTest, mapStateToProps } from '../../../src/app/components/ScenariosDisplay';
import useScenarios from '../../../src/app/hooks/useScenarios';

jest.mock('../../../src/app/hooks/useScenarios', () => ({
  __esModule: true,
  default: jest.fn(() => ({
    loading: false,
    scenario: '',
    filteredScenarios: [],
    filterValue: '',
    setFilterValue: jest.fn(),
    setScenario: jest.fn(() => Promise.resolve()),
    loadScenarios: jest.fn(() => Promise.resolve()),
  })),
}));

describe('ScenariosDisplay', () => {
  it('should mapStateToProps', () => {
    expect(mapStateToProps({ url: 'squawk' })).toMatchObject({ url: 'squawk' });
  });

  describe('ScenariosDisplay', () => {
    const url = 'https://dev.example.com/squawk';

    it('should render ScenariosDisplay component', () => {
      const result = shallow(<ComponentUnderTest url={url} />);

      expect(useScenarios).toHaveBeenCalledTimes(1);
      expect(useScenarios).toHaveBeenCalledWith(url);
      expect(result).toMatchSnapshot();
    });
  });
});
