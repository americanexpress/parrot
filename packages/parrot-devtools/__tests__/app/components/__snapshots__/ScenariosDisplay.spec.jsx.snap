// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScenariosDisplay ScenariosDisplay should render ScenariosDisplay component 1`] = `
<Fragment>
  <Toolbar
    scenariosData={
      Object {
        "filterValue": "",
        "filteredScenarios": Array [],
        "loadScenarios": [MockFunction],
        "loading": false,
        "scenario": "",
        "setFilterValue": [MockFunction],
        "setScenario": [MockFunction],
      }
    }
  />
  <styled.div>
    <ScenariosMiddleware
      scenariosData={
        Object {
          "filterValue": "",
          "filteredScenarios": Array [],
          "loadScenarios": [MockFunction],
          "loading": false,
          "scenario": "",
          "setFilterValue": [MockFunction],
          "setScenario": [MockFunction],
        }
      }
    />
  </styled.div>
</Fragment>
`;
